src/query/interpreter.cpp

```diff
@@ -1909,6 +1918,119 @@ PreparedQuery PrepareSettingQuery(ParsedQuery parsed_query, const bool in_explic
+                                     DbAccessor *db_accessor) {
+  Frame frame(0);
+  SymbolTable symbol_table;
+  EvaluationContext evaluation_context;
+  evaluation_context.timestamp = QueryTimestamp();
+  evaluation_context.parameters = parameters;
+  ExpressionEvaluator evaluator(&frame, symbol_table, evaluation_context, db_accessor, storage::View::OLD);
+
+  bool isAdmin = interpreter_context->auth_checker->IsUserAdmin(username);
+
+  Callback callback;
+  switch (transaction_query->action_) {
+    case TransactionQueueQuery::Action::SHOW_TRANSACTIONS: {
+      callback.header = {"username", "transaction_id", "queries summary"};
+      callback.fn = [interpreter_context, username, isAdmin]() mutable {
+        std::vector<std::vector<TypedValue>> results;
+        results.reserve(interpreter_context->interpreters->size());
+        interpreter_context->interpreters.WithLock([&results, username, isAdmin](const auto &interpreters) {
+          spdlog::info("Number of interpreters in callback: {}", interpreters.size());
+          for (Interpreter *interpreter : interpreters) {
+            if (interpreter->HasActiveTransaction() && (interpreter->username_ == username || isAdmin)) {
+              auto queries_summaries_typed = std::vector<TypedValue>();
+              auto queries_summaries = interpreter->GetQueriesSummary();  // in std::transform
+              std::for_each(queries_summaries.begin(), queries_summaries.end(),
+                            [&queries_summaries_typed](const auto &query_summary) {
+                              queries_summaries_typed.emplace_back(query_summary);
+                            });
+
+              results.push_back({TypedValue(interpreter->username_.value_or("")),
+                                 TypedValue(std::to_string(interpreter->GetTransactionId())),
+                                 TypedValue(queries_summaries_typed)});
+            }
+          }
+        });
+        return results;
+      };
+      break;
+    }
+    case TransactionQueueQuery::Action::TERMINATE_TRANSACTIONS: {
+      std::vector<std::string> transaction_ids;
+      std::transform(transaction_query->transaction_id_list_.begin(), transaction_query->transaction_id_list_.end(),
+                     std::back_inserter(transaction_ids), [&evaluator](Expression *expression) {
+                       return std::string(expression->Accept(evaluator).ValueString());
+                     });
+      callback.header = {"transaction_id", "status"};
+      callback.fn = [interpreter_context, db_accessor, transaction_ids, username, isAdmin]() mutable {
+        spdlog::info("Number of transactions to kill in callback: {}", transaction_ids.size());
+        std::vector<std::vector<TypedValue>> results;
+        for (const std::string &transaction_id : transaction_ids) {
+          std::string status{"Transaction not found"};
+          interpreter_context->interpreters.WithLock(
```

Comment:
I think we can make this more readable and I have two ideas, so you can choose if you think it will work:

1. Instead of going over all interpreters for each transaction id, I think you could do following.

On `interpreter_contex->interpreters` call  `Lock()`.

Like this:
```
auto my_map_ptr = my_important_map_.Lock();
my_map_ptr->emplace(key, value);
```

Then filter out interpreters which have active transaction and where transaction id is std::vector<std::string> (you can change this to std::unordered_set) and username is correct or isAdmin.

And afterwards just kill those transactions.

2. idea is following:
for (Interpreter *interpreter : interpreters) {
    if (!interpreter->HasActiveTransaction() ||  !transaction_id == std::to_string(interpreter->GetTransactionId())) {
        continue;
    }
    if (interpreter->username_ == username || isAdmin){
        status = "Not enough rights to kill the transaction";
        break;
    }
     
    
    interpreter->AbortTransactionByUser();
    interpreter->Abort();
    status = "Transaction killed";
    break;
}

Reply:
yes but you still have to iterate over transaction_ids sent by user, otherwise which ones you kill? I used this approach of locking inside the loop of transaction_ids so that interpreters don't stay locked for longer time than neccessary